src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioSymbolNavigationService.cs

```diff
@@ -115,14 +115,22 @@ public bool TryNavigateToSymbol(ISymbol symbol, Project project, OptionSet? opti
             }
 
             // Generate new source or retrieve existing source for the symbol in question
+            return ThreadingContext.JoinableTaskFactory.Run(() => TryNavigateToMetadataAsync(project, symbol, options, cancellationToken));
+        }
+
+        private async Task<bool> TryNavigateToMetadataAsync(Project project, ISymbol symbol, OptionSet options, CancellationToken cancellationToken)
+        {
             var allowDecompilation = _globalOptions.GetOption(FeatureOnOffOptions.NavigateToDecompiledSources);
-            var result = ThreadingContext.JoinableTaskFactory.Run(() => _metadataAsSourceFileService.GetGeneratedFileAsync(project, symbol, signaturesOnly: false, allowDecompilation, cancellationToken));
+
+            var result = await _metadataAsSourceFileService.GetGeneratedFileAsync(project, symbol, signaturesOnly: false, allowDecompilation, cancellationToken).ConfigureAwait(false);
+
+            await this.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
nit: can we just `ConfigureAwait(true)` here? Or is that too subtle?

Reply:
Do you mean on line 125? That won't help, as we're not on the UI thread before that line (and I suspect it used to be, hence the bug).

If you mean on the `SwitchToMainThreadAsync` call, I don't think I've ever seen a `ConfigureAwait` on them.

Reply:
I meant line 125. Isn't this on the UI thread because of `ThreadingContext.JoinableTaskFactory.Run`?

Reply:
nm, I have to read the documentation... every... time... for JTF

Reply:
I didn't think Run switched, I thought it just blocked